### PR TITLE
[scala3] Enable scala3_sources

### PR DESCRIPTION
**Pull request by kczulko**
This PR attempts to enable `scala3_sources` scalapb option which is [described here](https://scalapb.github.io/docs/customizations#file-level-options).

### CODE REVIEW

**Comment by pjfanning** [comment]
@kczulko will this break on our Scala 2.12 build?

**Comment by kczulko** [comment]
I don't think so. We're using both scala2 and scala3 in our project (event the same sbt module is being compiled twice, due to porting-to-scala3 process...). I've been checking this via `+publishLocal` and all was fine. Also tests were passed.

**Comment by mdedetrich** [comment]
My gut feeling is that this should be targeting 1.1.x, @pjfanning wdyt?

**Comment by pjfanning** [comment]
it doesn't yet work - it seems to break Scala 2.12 as I suspected it might https://github.com/apache/incubator-pekko-grpc/actions/runs/7797963372/job/21265803687?pr=222

**Comment by mdedetrich** [comment]
> it doesn't yet work - it seems to break Scala 2.12 as I suspected it might https://github.com/apache/incubator-pekko-grpc/actions/runs/7797963372/job/21265803687?pr=222

Probably need to pass some flags to `scalacOptions` so that the scala 2.12 compiler can accept the scala 3 source style

**Comment by kczulko** [comment]
> Probably need to pass some flags to scalacOptions so that the scala 2.12 compiler can accept the scala 3 source style

The errors are coming from java files. For now I don't get it what could cause this issue. Do you have some suggestions?

**Comment by kczulko** [comment]
As I'm bumping the scalapb compilerplugin I suspect the incompatibility may be here? https://github.com/scalapb/ScalaPB/blob/71befffd3db412d158fabb807ded56920fa0030f/project/Dependencies.scala#L7

**Comment by pjfanning** [comment]
yes - a big bump in io.grpc libs causes us issues - eg https://github.com/apache/incubator-pekko-grpc/pull/190

**Comment by kczulko** [comment]
@pjfanning 
Regarding failed test case... There's the [protobuf-javascript issue](https://github.com/protocolbuffers/protobuf-javascript/issues/127). Do want to workaround it somehow? It would require more work to get fixed. E.g. as a nixos user I added `protoc-gen-go` to my nix shell (so that it appears in PATH) and slightly modified 06-compatibility-plugins test case. With nix it's easy however if you'd like to avoid nix, then it required more code to determine developer's OS and whole download/unzip/(add-to-path | configure-build.sbt) logic.

Next thing - I don't know why `Link Validator` build step failed :)

**Comment by pjfanning** [comment]
The Link Validator tests are very brittle. It calls all the linked sites and many of them fail on occasion.
The compatibility test is something we can probably rework or remove. Let me discuss it with other contributors.

**Comment by kczulko** [comment]
@pjfanning @mdedetrich 
Please let me know if there's anything you want from me to push it forward.

Thanks in advance,
Karol

**Comment by raboof** [comment]
> With nix it's easy however if you'd like to avoid nix, then it requires more code to determine developer's OS and whole download/unzip/(add-to-path | configure-build.sbt) logic. Or some shenanigans could be done on the .github workflow files but I would rather want to keep coherent configuration for both local and CI envs

since this is an sbt integration test, I think it's OK if there's some prerequisites for running it (such as putting `protoc-gen-go` on the PATH). If we can add that to the GHA script and document that the user needs to do that themselves (however it needs to be done on their OS), then I think we're OK.

**Comment by kczulko** [comment]
@raboof 

Ok then. Let me try to do it. I'll use [this gh action](https://github.com/actions/setup-go) together with [this go install command](https://protobuf.dev/getting-started/gotutorial/#compiling-protocol-buffers). However, I'm not allowed to check how this will work under github actions env since contributors have to wait for the CI run approval. Therefore I'd like to ask you (the contributors) to eventually fix all the shortcomings (as you can probably do it without waiting for approvals).

Best regards,
Karol

File: codegen/src/main/scala/org/apache/pekko/grpc/gen/scaladsl/ScalaCompatConstants.scala
Comment (raboof) [inline codegen/src/main/scala/org/apache/pekko/grpc/gen/scaladsl/ScalaCompatConstants.scala:21]: I guess this can be dropped

File: runtime/src/main/mima-filters/1.1.x.backwards.excludes/io.grpc-upgrade.backwards.excludes
Comment (raboof) [inline runtime/src/main/mima-filters/1.1.x.backwards.excludes/io.grpc-upgrade.backwards.excludes:19]: :+1: since internal

File: interop-tests/src/test/scala/org/apache/pekko/grpc/scaladsl/NonBalancingIntegrationSpec.scala
Comment (raboof) [inline interop-tests/src/test/scala/org/apache/pekko/grpc/scaladsl/NonBalancingIntegrationSpec.scala:186]: strange but no problem

**Comment by raboof** [comment]
Great work here!

I'm confused why the `Validate and test` CI workflow (from `build-test.yml`) isn't running here

File: sbt-plugin/src/sbt-test/gen-scala-server/06-compatibility-plugins/project/ProtocJSPlugin.scala
Comment (samueleresca) [inline sbt-plugin/src/sbt-test/gen-scala-server/06-compatibility-plugins/project/ProtocJSPlugin.scala:24]: Should the `ProtocJSPlugin` object renamed to something `ProtocGoPlugin`?

**Comment by raboof** [comment]
the GHA error seems to be 'every step must define a `uses` or `run` key' - though this is strange since it seems you do.

File: runtime/src/main/scala/org/apache/pekko/grpc/internal/PekkoDiscoveryNameResolverProvider.scala
Comment (kczulko) [reply runtime/src/main/scala/org/apache/pekko/grpc/internal/PekkoDiscoveryNameResolverProvider.scala:39]: It looks like `authority` is no longer passed when `newNameResolver` is called.

**Comment by samueleresca** [comment]
> the GHA error seems to be 'every step must define a `uses` or `run` key' - though this is strange since it seems you do.

I think that error is caused by the `Install go & go-protobuf` step defined with 2 "-":

```
      - name: Install go & go-protobuf
      - uses: actions/setup-go@v5
      ...
```

File: .github/workflows/build-test.yml
Comment (raboof) [inline .github/workflows/build-test.yml:118]: Ah looks like we have to split this into two: "a step cannot have both the `uses` and `run` keys"

**Comment by kczulko** [comment]
Dear Lord... I think it's ok now :D

**Comment by raboof** [comment]
It seems the link validator isn't wrong here, and the license report is linking to http://www.jetbrains.org for a new `org.jetbrains:annotations:13.0` dependency. We might want to add http://www.jetbrains.org to the `non-https-whitelist` in `./scripts/link-validator.conf`?

**Review by raboof — APPROVED** [review]

**Pull request merged**
